src/osiris_competing_read_coordinator.erl

```diff
@@ -0,0 +1,334 @@
+%% @end
+%%--------------------------------------------------------------------
+-spec init(Args :: term()) ->
+              {ok, State :: term()} |
+              {ok, State :: term(), Timeout :: timeout()} |
+              {ok, State :: term(), hibernate} |
+              {stop, Reason :: term()} |
+              ignore.
+init([Config, SupPid]) ->
+    process_flag(trap_exit, true),
+    gen_server:cast(self(), init),
+    {ok,
+     #state{config = Config,
+            sup = SupPid,
+            readers = #{},
+            reader_groups = #{},
+            service_queue = queue:new(),
+            pending = [],
+            batch_size =
+                application:get_env(osiris, crc_batch_size,
+                                    ?DEFAULT_BATCH_SIZE)}}.
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% Handling call messages
+%% @end
+%%--------------------------------------------------------------------
+-spec handle_call(Request :: term(), From :: {pid(), term()},
+                  State :: term()) ->
+                     {reply, Reply :: term(), NewState :: term()} |
+                     {reply,
+                      Reply :: term(),
+                      NewState :: term(),
+                      Timeout :: timeout()} |
+                     {reply, Reply :: term(), NewState :: term(), hibernate} |
+                     {noreply, NewState :: term()} |
+                     {noreply, NewState :: term(), Timeout :: timeout()} |
+                     {noreply, NewState :: term(), hibernate} |
+                     {stop,
+                      Reason :: term(),
+                      Reply :: term(),
+                      NewState :: term()} |
+                     {stop, Reason :: term(), NewState :: term()}.
+handle_call({register, Pid, Tag, Prefetch, Fmt}, _From,
+            #state{readers = Readers,
+                   reader_groups = Groups0,
+                   service_queue = SQ} =
+                State) ->
+    Reader = {Pid, Tag},
+    case maps:get(Reader, Readers, undefined) of
+        undefined ->
+            ReaderState =
+                #{prefetch => Prefetch,
+                  formatter => Fmt,
+                  credit => Prefetch,
+                  in_flight => gb_sets:new()},
+            Groups =
+                case maps:get(Pid, Groups0, undefined) of
+                    undefined ->
+                        MRef = monitor(process, Pid),
+                        Groups0#{Pid => #{monitor_ref => MRef, tags => [Tag]}};
+                    #{tags := Tags} = Group ->
+                        Groups0#{Pid => Group#{tags => [Tag | Tags]}}
+                end,
+            {reply, ok,
+             notify_readers(State#state{readers =
+                                            maps:put(Reader, ReaderState,
+                                                     Readers),
+                                        service_queue = queue:in(Reader, SQ),
```

Comment:
I've got a feeling this will need to be a priority queue eventually... Not sure if we can extract the rabbit one or just grab something else